GridLocation gridLocation
MapLocation mapLocation
string state